src/Utilities/TrackedDependencies/FileTracker.cs

```diff
@@ -333,40 +303,26 @@ public static string FormatRootingMarker(ITaskItem[] sources)
         /// <param name="outputs">ITaskItem array of outputs.</param>
         public static string FormatRootingMarker(ITaskItem[] sources, ITaskItem[] outputs)
         {
-            ErrorUtilities.VerifyThrowArgumentNull(sources, "sources");
+            ErrorUtilities.VerifyThrowArgumentNull(sources, nameof(sources));
 
-            ArrayList rootSources = new ArrayList();
-            StringBuilder rootSourcesList = new StringBuilder();
-            int builderLength = 0;
+            outputs = outputs ?? Array.Empty<ITaskItem>();
 
-            foreach (ITaskItem source in sources)
-            {
-                rootSources.Add(FileUtilities.NormalizePath(source.ItemSpec).ToUpperInvariant());
-            }
+            var rootSources = new string[sources.Length + outputs.Length];
 
-            if (outputs != null)
+            var i = 0;
+            for (; i < sources.Length; i++)
             {
-                foreach (ITaskItem output in outputs)
-                {
-                    rootSources.Add(FileUtilities.NormalizePath(output.ItemSpec).ToUpperInvariant());
-                }
+                rootSources[i] = FileUtilities.NormalizePath(sources[i].ItemSpec).ToUpperInvariant();
             }
 
-            rootSources.Sort(StringComparer.OrdinalIgnoreCase);
-
-            foreach (string source in rootSources)
+            for (var j = 0; j < outputs.Length; j++, i++)
             {
-                rootSourcesList.Append(source);
-                rootSourcesList.Append('|');
+                rootSources[i] = FileUtilities.NormalizePath(outputs[j].ItemSpec).ToUpperInvariant();
```

Comment:
I'm incrementing `i` as well, so just `i` is correct. Alternately I could stop incrementing `i` and do `i + j`. I wasn't sure which was cleaner to others.

---
In reply to: [197191491](https://github.com/Microsoft/msbuild/pull/3438#discussion_r197191491) [](ancestors = 197191491)

Reply:
Pre-allocated list should be about the same (I assume `List.Sort` and `string.Join` work just as far on Lists as they do arrays. You're right that it would look nicer as well.

---
In reply to: [197197005](https://github.com/Microsoft/msbuild/pull/3438#discussion_r197197005) [](ancestors = 197197005)